packages/server/src/plugin/subscriptionCallback/index.ts

```diff
@@ -0,0 +1,672 @@
+        // reason to let the interval continue to run.
+        clearInterval(heartbeatInterval);
+        this.logger?.error(
+          `Programming error: Heartbeat interval unexpectedly missing for ${callbackUrl}. This is probably a bug in Apollo Server.`,
+        );
+        return;
+      }
+      const { queue } = existingHeartbeat;
+      queue.push(heartbeatPromise);
+      if (queue.length > 1) {
+        const requestBeforeMe = queue[existingHeartbeat?.queue.length - 2];
+        await requestBeforeMe;
+      }
+
+      // Send the heartbeat request
+      try {
+        const ids = Array.from(
+          this.subscriptionInfoByCallbackUrl
+            .get(callbackUrl)
+            ?.subscriptionsById.keys() ?? [],
+        );
+        this.logger?.debug(
+          `Sending \`heartbeat\` request to ${callbackUrl} for IDs: [${ids.join(
+            ',',
+          )}]`,
+        );
+
+        heartbeatRequest = fetch(callbackUrl, {
+          method: 'POST',
+          body: JSON.stringify({
+            kind: 'subscription',
+            action: 'heartbeat',
+            id: existingHeartbeat?.id ?? id,
+            verifier: existingHeartbeat?.verifier ?? verifier,
+            ids,
+          }),
+          headers: { 'Content-Type': 'application/json' },
+        });
+        this.requestsInFlight.add(heartbeatRequest);
+
+        // The heartbeat response might contain updates for us to act upon, so we
+        // need to await it
+        const result = await heartbeatRequest;
+        consecutiveHeartbeatFailureCount = 0;
+
+        this.logger?.debug(
+          `Heartbeat request successful for IDs: [${ids.join(',')}]`,
+        );
+
+        if (result.status === 400) {
+          const body = await result.json();
+          this.logger?.debug(
+            `Heartbeat request received invalid IDs: [${body.invalid_ids.join(
+              ',',
+            )}]`,
+          );
+          // Some of the IDs are invalid, so we need to update the id and
+          // verifier for future heartbeat requests (both provided by the router
+          // in the response body)
+          existingHeartbeat.id = body.id;
+          existingHeartbeat.verifier = body.verifier;
+
+          this.terminateSubscriptions(body.invalid_ids, callbackUrl);
+        } else if (result.status === 404) {
+          // all ids we sent are invalid
+          this.logger?.debug(
+            `Heartbeat request received invalid IDs: [${ids.join(',')}]`,
+          );
+          // This will also handle cleaning up the heartbeat interval
+          this.terminateSubscriptions(ids, callbackUrl);
+        }
+      } catch (e) {
+        // The heartbeat request failed.
+        this.logger?.error(
+          `Heartbeat request failed (${++consecutiveHeartbeatFailureCount} consecutive): ${e}`,
+          existingHeartbeat.id,
+        );
+
+        if (consecutiveHeartbeatFailureCount >= 5) {
+          this.logger?.error(
```

Comment:
Do you need to clearInterval in this block?

Reply:
Handled by the call to `this.terminateSubscriptions` immediately below and noted in the comment